eth/tracers/live/live.go

```diff
@@ -0,0 +1,364 @@
+}
+
+const (
+	tableSize = 2 * 1024 * 1024 * 1024
+)
+
+type traceResult struct {
+	TxHash *common.Hash `json:"txHash,omitempty"` // Transaction hash generated from block
+	Result interface{}  `json:"result,omitempty"` // Trace results produced by the tracer
+	Error  string       `json:"error,omitempty"`  // Trace failure produced by the tracer
+}
+
+// EncodeRLP implments rlp.Encoder
+func (tr *traceResult) EncodeRLP(w io.Writer) error {
+	return rlp.Encode(w, []interface{}{tr.Result, tr.Error})
+}
+
+// DecodeRLP implements rlp.Decoder
+func (tr *traceResult) DecodeRLP(s *rlp.Stream) error {
+	var temp struct {
+		Result []byte
+		Error  string
+	}
+	if err := s.Decode(&temp); err != nil {
+		return err
+	}
+	tr.Error = temp.Error
+	return json.Unmarshal(temp.Result, &tr.Result)
+}
+
+type live struct {
+	backend   tracing.Backend
+	kvdb      ethdb.Database
+	frdb      *rawdb.Freezer
+	freezeCh  chan uint64
+	stopCh    chan struct{}
+	tables    map[string]bool
+	traces    map[string][]*traceResult
+	tracer    *native.MuxTracer
+	latest    atomic.Uint64
+	offset    atomic.Uint64
+	finalized uint64
+	hash      common.Hash
+	once      sync.Once
+
+	enableNonceTracer bool
+}
+
+type liveTracerConfig struct {
+	Path              string          `json:"path"` // Path to the directory where the tracer data will be stored
+	Config            json.RawMessage `json:"config"`
+	EnableNonceTracer bool            `json:"enableNonceTracer"`
+	MaxKeepBlocks     uint64          `json:"maxKeepBlocks"` // Maximum number of blocks to keep in the freezer db(the unconfirmaed blocks are not included), 0 means no limit
+}
+
+func toTraceTable(name string) string {
+	return name + "_traces"
+}
+
+// encodeNumber encodes a number as big endian uint64
+func encodeNumber(number uint64) []byte {
+	enc := make([]byte, 8)
+	binary.BigEndian.PutUint64(enc, number)
+	return enc
+}
+
+func toKVKey(name string, number uint64, hash common.Hash) []byte {
+	var typo byte
+	switch name {
+	case "callTracer":
+		typo = byte('C')
+	case "flatCallTracer":
+		typo = byte('F')
+	case "prestateTracer":
+		typo = byte('S')
+	case "parityTracer":
+		typo = byte('P')
+	default:
+		panic("not supported yet")
+	}
+	// TODO: have some prefix?
+	key := append(encodeNumber(number), hash.Bytes()...)
+	key = append(key, typo)
+
+	return key
+}
+
+func newLive(cfg json.RawMessage, stack tracers.LiveApiRegister, backend tracing.Backend) (*tracing.Hooks, error) {
+	var config liveTracerConfig
+	if cfg != nil {
+		if err := json.Unmarshal(cfg, &config); err != nil {
+			return nil, fmt.Errorf("failed to parse config: %v", err)
+		}
+	}
+	if config.Path == "" {
+		return nil, errors.New("live tracer output path is required")
+	}
+
+	t, err := native.NewMuxTracer(config.Config)
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		kvpath = path.Join(config.Path, "kvdb")
+		frpath = path.Join(config.Path, "frdb")
+	)
+
+	kvdb, err := rawdb.NewPebbleDBDatabase(kvpath, 128, 1024, "trace", false, false)
```

Comment:
Btw now that we are passing in the *node.Node handle we can use that to access databases.

Reply:
Yes, but I don't want to put the temporary data into the chaindata, as the data we used here is temporary, and acts in the order: write-once, read-once, deleted(level merged), so the user can set a separate, and cheaper SSD for such temporary data needs.